tests/unit/s2n_client_hello_get_supported_groups_test.c

```diff
@@ -0,0 +1,390 @@
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include "pq-crypto/s2n_pq.h"
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/extensions/s2n_client_supported_groups.h"
+#include "tls/s2n_client_hello.h"
+#include "tls/s2n_tls.h"
+#include "utils/s2n_random.h"
+
+#define S2N_TEST_SUPPORTED_GROUPS_LIST_COUNT 16
+
+/* Each supported group is 2 bytes. */
+#define S2N_TEST_SUPPORTED_GROUPS_LIST_SIZE (S2N_TEST_SUPPORTED_GROUPS_LIST_COUNT * S2N_SUPPORTED_GROUP_SIZE)
+
+/* 2 length bytes + space for the list of supported groups. */
+#define S2N_TEST_SUPPORTED_GROUPS_EXTENSION_SIZE (2 + S2N_TEST_SUPPORTED_GROUPS_LIST_SIZE)
+
+struct s2n_client_hello_context {
+    struct s2n_stuffer *sent_supported_groups_extension;
+    int invoked_count;
+};
+
+int s2n_check_received_supported_groups_cb(struct s2n_connection *conn, void *ctx)
+{
+    EXPECT_NOT_NULL(ctx);
+
+    struct s2n_client_hello_context *context = (struct s2n_client_hello_context *) ctx;
+    EXPECT_NOT_NULL(context->sent_supported_groups_extension);
+    context->invoked_count += 1;
+
+    struct s2n_client_hello *client_hello = s2n_connection_get_client_hello(conn);
+    EXPECT_NOT_NULL(client_hello);
+
+    uint16_t received_groups[S2N_TEST_SUPPORTED_GROUPS_LIST_COUNT] = { 0 };
+    uint16_t received_groups_count = 0;
+    EXPECT_SUCCESS(s2n_client_hello_get_supported_groups(client_hello, received_groups,
+            s2n_array_len(received_groups), &received_groups_count));
+
+    uint16_t sent_groups_count = 0;
+    EXPECT_OK(s2n_supported_groups_parse_count(context->sent_supported_groups_extension, &sent_groups_count));
+    EXPECT_EQUAL(received_groups_count, sent_groups_count);
+
+    for (size_t i = 0; i < received_groups_count; i++) {
+        uint16_t received_group = received_groups[i];
+
+        uint16_t sent_group = 0;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint16(context->sent_supported_groups_extension, &sent_group));
+
+        EXPECT_EQUAL(received_group, sent_group);
+    }
+
+    return S2N_SUCCESS;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key *chain_and_key = NULL, s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    /* Safety */
+    {
+        struct s2n_client_hello client_hello = { 0 };
+        uint16_t supported_groups[S2N_TEST_SUPPORTED_GROUPS_LIST_COUNT] = { 0 };
+        uint16_t supported_groups_count = 0;
+
+        int ret = s2n_client_hello_get_supported_groups(NULL, supported_groups, s2n_array_len(supported_groups),
+                &supported_groups_count);
+        EXPECT_FAILURE_WITH_ERRNO(ret, S2N_ERR_NULL);
```

Comment:
```suggestion
        EXPECT_FAILURE_WITH_ERRNO( s2n_client_hello_get_supported_groups(NULL, supported_groups, s2n_array_len(supported_groups), &supported_groups_count), S2N_ERR_NULL);
```
Is there a reason why you can't do this? Am I missing something? You seem to be doing this in a lot of places in this test.

Reply:
Clang format https://github.com/aws/s2n-tls/pull/4216#discussion_r1333511930